shinysdr/plugins/limesdr.py

```diff
@@ -0,0 +1,317 @@
+        name=None,
+        calibration=True):
+    """
+    serial: device serial number
+    device_type: LimeSDRMini or LimeSDRUSB
+    lna_path: LNANONE, LNAL, LNAH, or LNAW
+    name: block name (usually not specified)
+    sample_rate: desired sample rate, or None == guess a good rate
+    
+    See documentation in shinysdr/i/webstatic/manual/configuration.html.
+    """
+    
+    if not _available:
+        raise Exception('LimeSDRDevice: gr-limesdr Python bindings not found; cannot create device')
+    
+    serial = defaultstr(serial)
+    if name is None:
+        name = 'LimeSDR %s' % serial
+
+    # TODO: High gain might be unsafe, but low gain might result in failed calibration.
+    # Ideally we'd load these initial values from the saved state?
+    freq = 1e9
+    gain = 50
+    if_bandwidth = 3e6
+
+    source = create_source(serial, device_type, lna_path, sample_rate, freq, if_bandwidth, gain, calibration=calibration)
+    
+    tuning = _LimeSDRTuning(source)
+    vfo_cell = tuning.get_vfo_cell()
+    
+    rx_driver = _LimeSDRRXDriver(
+        device_type=device_type,
+        source=source,
+        lna_path=lna_path,
+        name=name,
+        tuning=tuning,
+        sample_rate=sample_rate)
+    
+    vfo_cell.set(freq)
+    
+    return Device(
+        name=name,
+        vfo_cell=vfo_cell,
+        rx_driver=rx_driver,
+        tx_driver=nullExportedState)
+
+
+__all__.append('LimeSDRDevice')
+
+
+@implementer(IRXDriver)
+class _LimeSDRRXDriver(ExportedState, gr.hier_block2):
+    
+    # Note: Docs for gr-limesdr are in comments at gr-limesdr/include/limesdr/source.h
+    def __init__(self,
+                 source,
+                 device_type,
+                 lna_path,
+                 name,
+                 tuning,
+                 sample_rate):
+        gr.hier_block2.__init__(
+            self, defaultstr('RX ' + name),
+            gr.io_signature(0, 0, 0),
+            gr.io_signature(1, 1, gr.sizeof_gr_complex * 1),
+        )
+        
+        self.__source = source
+        self.__name = name
+        self.__tuning = tuning
+        
+        self.connect(self.__source, self)
+        
+        # State of the source that there are no getters for, so we must keep our own copy of
+        self.__track_gain = 50.
+        source.set_gain(int(self.__track_gain), ch)
+
+        self.__track_bandwidth = max(sample_rate / 2, 1.5e6)
+        source.set_analog_filter(True, self.__track_bandwidth, ch)
+
+        self.__lna_path_type = EnumT({
+            LNANONE: 'None',
+            LNAH: 'LNAH',
+            LNAL: 'LNAL',
+            LNAW: 'LNAW',
+        })
+        if device_type == LimeSDRMini:
+            self.__lna_path_type = EnumT({
+                LNAH: 'LNAH',
+                LNAW: 'LNAW',
+            })
+        self.__track_lna_path = lna_path
+        
+        # Blocks
+        self.__state_while_inactive = {}
+        self.__placeholder = blocks.vector_source_c([])
```

Comment:
`__state_while_inactive` and `__placeholder` are both only needed for the mutually exclusive source and sink logic in OsmoSDRDevice.

Reply:
Done.